File: code/java/剑指offer48.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int res = 0;
        Map<Character,Integer> map = new HashMap<>();
        int i = 0;
        int j = 0;
        while(j<s.length()){
            char c = s.charAt(j);
            if(!map.containsKey(c)){
                map.put(c,1);
                res = Math.max(res,j-i+1);
                j+=1;
            }
            else if(map.get(c)==0){
                map.remove(c);
                map.put(c,1);
                res = Math.max(res,j-i+1);
                j+=1;
            }
            else{
                while(i<j){
                    if(s.charAt(i)==s.charAt(j)){
                        map.remove(s.charAt(j));
                        break;
                    }
                    int temp = map.get(s.charAt(i));
                    map.remove(s.charAt(i));
                    map.put(s.charAt(i),temp-1);
                    i++;
                }
                i++;
            }
        }
        res = Math.max(res,j-i);
        return res;
    }
}

File: code/python/剑指offer48.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        curMap = {}
        i = 0
        j = 0
        while j<len(s):
            c = s[j]
            if c not in curMap or curMap[c]==0:
                curMap[c] = 1
                res = max(res,j-i+1)
                j+=1
            else:
                while True and i<j:
                    if s[i]==s[j]:
                        curMap[s[j]]-=1
                        break
                    curMap[s[i]]-=1
                    i+=1
                i+=1
        res = max(res,j-i)
        return res